docs/experiments-vector-stores.md

```diff
@@ -0,0 +1,286 @@
+```
+
+## DuckDB
+Let's start with DuckDB. Install it with:
+```bash
+pip install duckdb
+```
+
+Then, we can connect to the database in Python. 
+```python
+import duckdb
+conn = duckdb.connect(":memory:")
+```
+
+Now, we initialize and load tables for our corpus and queries. We use DuckDB's float array to hold our embeddings.
+```python
+corpus_path = 'indexes/nfcorpus.bge-base-en-v1.5/corpus_embeddings.jsonl'
+query_path = 'indexes/nfcorpus.bge-base-en-v1.5/query_embeddings.jsonl'
+
+embd_dim = 0
+import json
+with open(corpus_path, 'r') as file:
+    for line in file:
+        row = json.loads(line.strip())
+        embd_dim = len(row['vector'])
+        break
+
+conn.execute(f"""create table corpus (id varchar primary key, embedding float[{embd_dim}])""")
+conn.execute(f"""create table query (id varchar primary key, embedding float[{embd_dim}])""")
+
+def load_jsonl_to_table(file_path, table_name):
+    with open(file_path, 'r') as file:
+        for line in file:
+            row = json.loads(line.strip())
+            a = conn.execute(f"""insert into {table_name} (id, embedding) values (?, ?)""", (row['id'], row['vector']))
+
+load_jsonl_to_table(corpus_path, "corpus")
+load_jsonl_to_table(query_path, "query")
+```
+
+Let's define a method for retrieving results for one query. We obtain the query embeddings with the query ID passed in and use DuckDB's ```array_cosine_similarity``` method to find the closest document embeddings to our query embeddings. 
+```python
+def embedding_search(query_id, top_n=5):
+    query = f"""
+    WITH query_embedding AS (
+        SELECT embedding FROM query WHERE id = ?
+    )
+    SELECT corpus.id, 
+            array_cosine_similarity(corpus.embedding, query_embedding.embedding) AS score
+    FROM corpus, query_embedding
+    ORDER BY score DESC
+    LIMIT ?
+    """
+    return conn.execute(query, [query_id, top_n]).fetchall()
+```
+
+We call the retrieval method on all our queries to retrieve the top 1000 results for each. 
+```python
+from tqdm import tqdm
+queries = conn.execute("SELECT id, contents FROM query").fetchall()
+run_tag = "bge_duckdb"
+
+all_results = []
+
+for query_id, query_string in tqdm(queries, desc=f"Processing {run_tag}", unit="query"):
+    results = embedding_search(query_id, top_n=1000)
+    for rank, (doc_id, score) in enumerate(results, 1):
+        all_results.append((query_id, doc_id, score, rank))
+
+all_results.sort(key=lambda x: (x[0], x[3])) # sort by queryid, then rank
+
+with open("runs/duckdb_bge_nfcorpus.txt", "w") as f:
+    for query_id, doc_id, score, rank in all_results:
+        a = f.write(f"{query_id} Q0 {doc_id} {rank} {score} {run_tag}\n")
+```
+
+To evaluate our results:
+```
+python -m pyserini.eval.trec_eval -c -m ndcg_cut.10 collections/nfcorpus/qrels/test.qrels runs/duckdb_bge_nfcorpus.txt
+```
+which should yield:
+
+| **Retrieval Method**                                                                                                  | **nDCG@10**  |
```

Comment:
Can you gather all the rows of the results table and move to the top? (as a summary)